flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriter.java

```diff
@@ -0,0 +1,101 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.channel;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.state.StateObject;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+/**
+ * Writes channel state during checkpoint/savepoint.
+ */
+@Internal
+public interface ChannelStateWriter extends AutoCloseable {
+
+	/**
+	 * Initiate write of channel state for the given checkpoint id.
+	 */
+	void start(long checkpointId);
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.consumer.InputChannel InputChannel}.
+	 * <p>Must be called after {@link #start} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer
+	 *                    It’s intended use is incremental snapshots
+	 *                    For the restored buffers it should be -1
+	 *                    For MVP it can always be -2
+	 *                    If no data is passed it is ignored
+	 */
+	void addInputData(long checkpointId, InputChannelInfo info, int startSeqNum, Buffer... data);
+
+	/**
+	 * Add in-flight buffers from the {@link org.apache.flink.runtime.io.network.partition.ResultSubpartition ResultSubpartition}.
+	 * <p>Must be called after {@link #start} and before {@link #finish(long)}.</p>
+	 * @param startSeqNum is the sequence number of the 1st passed buffer
+	 *                    It’s intended use is incremental snapshots
+	 *                    For the restored buffers it should be -1
+	 *                    For MVP it can always be -2
+	 *                    If no data is passed it is ignored
+	 */
+	void addOutputData(long checkpointId, ResultSubpartitionInfo info, int startSeqNum, Buffer... data);
+
+	/**
+	 * Finalize write of channel state for the given checkpoint id.
+	 * <p>Must be called after {@link #start}.</p>
+	 */
+	void finish(long checkpointId);
+
+	/**
+	 * <p>Must be called after {@link #finish}.</p>
+	 */
+	Future<Collection<StateObject>> getWriteCompletionFuture(long checkpointId);
```

Comment:
Could we merge that with finish?
Do you really want to use Future instead of CompletableFuture?

Reply:
I'm not sure if these calls will come from the same place.
I think this can be refined once we get closer to state injection implementation.

Reply:
I guess it probably happens in different places. Anyway, it might be no harm to separate the different semantics.

Reply:
Agree.
I will also change the contract slightly so this method should be called after `start()`.

Reply:
Ah yes, if it does not need to be called after finish, it will be much more usable and I agree it should be separate.